docs/tech-spike-write-ups/get-an-identity-oidc-oauth-with-ruby.md

```diff
@@ -0,0 +1,126 @@
+Custom claims to expose a user’s TRN
+
+Two more features were thought of for the spike: customising the OAuth consent screens, and using an upstream SSO provider like Auth0 (and later GOV.UK Accounts). These were not attempted in the spike in the interest of saving time, but they should be possible.
+
+While the spike was successful, a number of risks were identified with taking this approach.
+Risks identified
+Risk #1: Understanding OAuth v2.0 and OpenID Connect
+OAuth and OIDC are complex specifications. They were designed to meet the needs of large-scale identity providers, like Okta, Twitter, Facebook, Microsoft. As such, a lot of thought and design has been put into them, but they are chiefly used by the “tech giants” which have significant resources to throw at this problem.
+
+The specifications are flexible enough to meet our needs, but come with a significant learning curve for the team implementing the provider-side of the specification. That team most likely needs at least one OIDC expert to unblock and upskill the rest of the team.
+
+This risk is exacerbated if the team is likely to experience churn or attrition, and doubly so if the SSO integration isn’t “completely standard.” That is the case for the NPQ Registration TRNless work, which further increases the risk.
+Risk #2: Maintaining doorkeeper, openid_connect, and other ruby libraries
+To produce the spike, the following open source gems were used:
+
+(server) doorkeeper - an OAuth v2.0 provider
+(server) doorkeeper-openid_connect - an OpenID Connect extension for doorkeeper
+(client) omniauth - an authentication client
+(client) omniauth_openid_connect - an OAuth v2.0 and OpenID Connect extension for omniauth
+
+These gems depend on a number of things, but worth mentioning are the following:
+
+openid_connect - the only “certified” OpenID Connect implementation for Ruby
+swd - a simple web discovery client library, used by openid_connect
+
+While it was possible to eventually achieve the goals of the spike, a number of issues were identified while working with the gems:
+
+The documentation is poor
+Example: doorkeeper’s explanation page for what an “Access Grant” is is a blank document
+The libraries have hardcoded defaults that don’t allow for certain use cases
+Example: swd hardcodes HTTPS as the only protocol that it will allow for the discovery endpoint, which means it’s not trivial to develop this system locally, without resorting to monkey patching or forking
+The libraries are not maintained and actively looking for maintainers
+Example: doorkeeper-openid_connect [meta] Looking for maintainers! #89
+The libraries have counter-intuitive behaviours
+Example: the example configuration for omniauth_openid_connect does not include an `issuer` field, which leads to a cryptic SSL error. Debugging this required digging 3 layers deep into the dependencies to find the issue above, regarding the hardcoded HTTPS
```

Comment:
```suggestion
- The documentation is poor
  - Example: doorkeeper’s explanation page for what an “Access Grant” is is a blank document
- The libraries have hardcoded defaults that don’t allow for certain use cases
  - Example: swd hardcodes HTTPS as the only protocol that it will allow for the discovery endpoint, which means it’s not trivial to develop this system locally, without resorting to monkey patching or forking
- The libraries are not maintained and actively looking for maintainers
  - Example: doorkeeper-openid_connect [meta] Looking for maintainers! #89
- The libraries have counter-intuitive behaviours
  - Example: the example configuration for omniauth_openid_connect does not include an `issuer` field, which leads to a cryptic SSL error. Debugging this required digging 3 layers deep into the dependencies to find the issue above, regarding the hardcoded HTTPS
```